designs/simultaneous-edit/README.md

```diff
@@ -0,0 +1,121 @@
+---
+state: draft
+---
+
+# Simultaneous Flow Editing
+
+## Summary
+
+With current Node-RED, multiple clients can connect to the same runtime and can edit the same flow independently.  However, editing results by multiple clients are not synchronized, so conflicts of flows occur during deployment of the flow.  There is a feature to support merging flow conflicts referring node property values, but it is difficult for end users to merge node properties in proper way.
+
+![conflict-detection.png](/Users/nisiyama/src/Curry/Git/designs/designs/simultaneous-edit/images/conflict-detection.png)
+
+![review-change.png](/Users/nisiyama/src/Curry/Git/designs/designs/simultaneous-edit/images/review-change.png)
+
+Therefore, this design document proposes a new Node-RED editor feature that supports multiple clients to edit the same flow simultaneously.
+
+## Authors
+
+- Hiroyasu Nishiyama
+
+## Proposal
+
+Here, we consider the following two methods as candidates for simultaneous editing support of a Node-RED flow:
+
+1. Synchronize editing actions among clients,
+
+2. Allow locking a tab by a client to make single client can only edit the tab.
+
+### Method1: Edit Action Synchronization
+
+#### 1-a: Requirements
+
+- The flow edit result of a client is immediately reflected to the flow edited on workspaces of other clients.
```

Comment:
It is easy to compare this to the user experience on Google Docs. However this is one very important difference.

In Google Docs, you are all editing the document on the server. No one has to hit save - the changes are 'live'.

In Node-RED, nothing is saved in the runtime until someone presses the Deploy button.

That difference is very important as it raises a number of questions over what happens when a user hits Deploy. 

They will end up deploying their work and whatever the other user was in the middle of doing - with no guarantee the other user was ready to deploy their work. They could completely break things. This is one of the reasons I haven't tackled this yet - despite a few early experiments. And it is why the current concurrent editing model is based around when a user Deploys their work.

Reply:
As I wrote above, within a small team, I think it's appropriate to deal with this by an arrangement between members.
Our users have similar issues with team development on current Node-RED and they are covering the issue by operations.
However, there may be some support features that need to be improved, such as a feature for helping participating users agree on deploy.

Reply:
My worry is the feature makes the collaboration issues *harder* - because there is nothing built-in to help co-ordinate deploys. User's could get into a *real mess* if this feature allows User A to deploy User B's half-finished work. Even if it relies on co-ordination outside of Node-RED, accidents still happen.

Reply:
I understand such concerns.  I think that one guard is that this feature is opt-in feature.  So, we can expect team members understand the behavior using it.
As for the support function for this, for example, notification to the user who is editing and authorization from them may be useful.

Reply:
I think the design needs to carefully describe what the user workflows are so we can see what type of interaction is possible and what issues come up.